bootloader/src/pyi_utils.c

```diff
@@ -895,33 +927,51 @@ pyi_utils_create_child(const char *thisfile, const int argc, char *const argv[])
             VS("WARNING: Application is started by systemd socket,"
                "but we can't set proper LISTEN_PID on it.\n");
         }
-        execvp(thisfile, argv_pyi);
+        if (execvp(thisfile, argv_pyi) < 0) {
+            VS("Failed to exec: %s\n", strerror(errno));
+            return 1;
+        }
+        /* NOTREACHED */
     }
-    /* Parent code. */
-    else {
-        child_pid = pid;
 
-        /* Redirect termination signals received by parent to child process. */
-        signal(SIGINT, &_signal_handler);
-        signal(SIGKILL, &_signal_handler);
-        signal(SIGTERM, &_signal_handler);
-    }
+    /* From here to end-of-function is parent code (since the child exec'd). */
 
-    wait(&rc);
+    child_pid = pid;
+
+    /* Redirect all signals received by parent to child process. */
+    if (ignore_signals) {
+        VS("LOADER: Ignoring all signals in parent\n");
+    } else {
+        VS("LOADER: Registering signal handlers\n");
+    }
+    for (signum = 0; signum < num_signals; ++signum) {
+        if (signum == SIGCHLD || signum == SIGCLD) {
+            // don't mess with SIGCHLD/SIGCLD; it affects our ability
+            // to wait() for the child to exit
+            handlers[signum] = SIG_ERR;
+        } else {
+            handlers[signum] = signal(signum, handler);
+        }
+    }
 
-    /* Parent code. */
-    if (child_pid != 0) {
-        /* When child process exited, reset signal handlers to default values. */
-        signal(SIGINT, SIG_DFL);
-        signal(SIGKILL, SIG_DFL);
-        signal(SIGTERM, SIG_DFL);
+    if (wait(&rc) < 0) {
+        VS("LOADER: wait(): %s\n", strerror(errno));
+        return 1;
+    }
 
-        for (i = 0; i < argc_pyi; i++) {
-            free(argv_pyi[i]);
+    /* When child process exited, reset signal handlers to default values. */
+    VS("LOADER: Restoring signal handlers\n");
+    for (signum = 0; signum < num_signals; ++signum) {
+        if (handlers[signum] != SIG_ERR) {
+            signal(signum, handlers[signum]);
```

Comment:
Not setting the signal back to SIG_DFL (as the old code did) is worth a commit of its own. Thus when reading the commit messages one can easily spot this relevant change.

Reply:
@htgoebel It's pretty insignificant, IMO, particularly since the bootloader (parent) is about to exit after this. I'd be more inclined to switch back to using SIG_DFL.